openhtf/exe/phase_executor.py

```diff
@@ -167,22 +173,30 @@ def execute_phases(self, phases):
     """
     for phase in phases:
       while True:
-        outcome = self.execute_one_phase(phase)
+        outcome = self._execute_one_phase(phase)
         if outcome:
+          # We have to run the teardown_func *before* we yield the outcome,
+          # because yielding the outcome results in the state being finalized
+          # in the case of a terminal outcome.
+          if outcome.is_terminal and teardown_func:
+            self._execute_one_phase(teardown_func, output_record=False)
           yield outcome
-
-          # If there was an error, stop executing now.
-          if outcome.raised_exception:
-            return
-
+  
+          # We shouldn't keep executing if the outcome was terminal.
+          if outcome.is_terminal:
```

Comment:
ah, good idea, took me a minute to figure out what you were suggesting.. i did hesitate a bit to implement `is_terminal`, since it's largely a duplication of `TestState.SetStatusFromPhaseOutcome()`, but it's a good idea to put this check directly in there so we get an obvious failure place if they get out of sync.